### ISSUE TITLE
iis_app not actually converting to app

### ISSUE THREAD

**Issue by zanzan42** [chef-cookbooks/iis#316]
### Cookbook version
5.0.8
### Chef-client version
12.2

### Platform Details
AWS Microsoft Windows Server 2012 R2 Base

### Scenario:
Using the iis_app LWRP as part of my cookbook (along with the base iis::default recipe, and iis_site and iis_pool LWRPs as well), the resultant configuration in IIS doesn't actually add the app (even though the Chef run logs it as successful). The directory containing the app shows up under the associated site, but I still have to manually use the "Convert to application..." context menu in the UI to make it an app.

### Steps to Reproduce:
Create an IIS site (pointed to the default application pool, for example; iis_site successfully does this)
Create a separate application pool for the app I'll be adding (iis_pool successfully does this)
Create the application subdirectory under the site root (s3_file and aws_opsworks_archive successfully pulls a zip file from S3 and expands the application directory under the site root: c:\inetpub\siteRoot\appRoot, for example)
Execute the iis_app resource which adds an app and links it to the site and appropriate application pool previously created by iis_site and iis_pool

Example resource block (using anonymized attribute references for this example):

```ruby
  iis_app node['foo'][app'] do
    site_name node['foo']['site_name']
    path node['foo']['app']['path']
    application_pool node['foo']['app']['application_pool']
    physical_path node['foo']['app']['physical_path']
    enabled_protocols node['foo']['app']['enabled_protocols']
    action :add
  end
```

### Expected Result:
The application should already be configured as such in the IIS console, linked to its assigned application pool, etc.

### Actual Result:
Although the directory for the application shows up under the site (as you would expect since it's under the siteRoot), it is not an actual application, just a directory. You still have to use the "Convert to application..." contextual menu in order to associate it with the application pool. It's as if iis_app didn't execute at all.

Perhaps I'm missing something.

**Comment by EasyAsABC123** [comment]
can you give me the values of your variables? @zanzan42 
otherwise it is difficult to help, i'm going to run the default tests for win2012r2, should work

**Comment by zanzan42** [comment]
iis_app "appName" do
    site_name "siteName"
    path "/"
    application_pool "appPool"
    physical_path "c:\\inetpub\\siteRoot\\appRoot"
    enabled_protocols "http"
    action :add
  end

I can't give you some of the actual attribute values, but the above is approximate.

IIS and its features were already in place from recipe includes for iis, iis::mod_aspnet45 (and a few other mod_* recipes).

The physical path already exists by virtue of previous s3_file and aws_opsworks_archive resources.

Both "siteName" and "appPool" already exist by virtue of the previous iis_site and iis_pool resources.

**Comment by EasyAsABC123** [comment]
yup, working on updating chef test kitchen to run my test recipe for everyone which generates all of these items.

**Comment by zanzan42** [comment]
Also, if it's helpful, since we have to maintain Chef 12.2 compatibility across our dependency chain, here's the relevant berksfile contents:

cookbook 'ohai', '= 4.2.3'
cookbook 'windows', '= 1.44.3'
cookbook 'iis', '~> 5.0.7'

**Comment by iennae** [comment]
Hi @zanzan42, just a heads up that we recognize that this is an issue. We will be working on the fix, but it will be specifically a 12.7+ fix. You will be able to use the fix if in your wrapper cookbook you include the [compat_resource cookbook](https://supermarket.chef.io/cookbooks/compat_resource).

**Comment by zanzan42** [comment]
Thanks for the update. Unfortunately, 12.7+ won't help me, as the AWS OpsWorks Stacks product is Chef 12.2. That's one of the reasons why I had to go through a dependency chain validation exercise and make sure that everything was set to a version with 12.1+ compatibility. But good to know I found a real thing.

**Comment by EasyAsABC123** [comment]
@zanzan42 you should be able to still use it with the compat_resource cookbook as stated by @iennae

**Comment by zanzan42** [comment]
I'll give that a try. I already have compat_resource included by berkshelf from other cookbook dependencies, but that has not allowed me to run any other cookbook versions with Chef version requirements above 12.2. I presume that's because my wrapper cookbook must have its own depends statement on compat_resource, but given the nature of AWS OpsWorks Stacks I'm not sure whether the version compatibility checking that it does during a custom cookbook validation will pass in the first place.

**Comment by EasyAsABC123** [comment]
This should be resolved in iis v6.0.0 #322

**Comment by zanzan42** [comment]
As I suspected, AWS OpsWorks Stacks (which uses Chef 12.2 for Windows stacks), throws an error. This is despite my wrapper cookbook (which is calling iis_app) having compat_resource listed as a dependency.

The OpsWorks chef run error is this (which is typically what you'd see when a cookbook uses the newer style custom resource definition rather than the older LWRP):

[2017-04-07T17:48:54+00:00] ERROR: Running exception handlers 
[2017-04-07T17:48:54+00:00] ERROR: Exception handlers complete 
[2017-04-07T17:48:54+00:00] FATAL: Stacktrace dumped to c:/chef/runs/2c93403f-afae-4c00-ba35-1579d9b11645/local-mode-cache/cache/chef-stacktrace.out 
[2017-04-07T17:48:54+00:00] FATAL: NoMethodError: undefined method `property' for Chef::Resource::IisApp:Class 

The full stacktrace on the host is this:

Generated at 2017-04-07 17:48:54 +0000
NoMethodError: undefined method `property' for Chef::Resource::IisApp:Class
c:/chef/runs/2c93403f-afae-4c00-ba35-1579d9b11645/local-mode-cache/cache/cookbooks/iis/resources/app.rb:25:in `class_fr
om_file'
c:/opscode/chef/embedded/apps/chef/lib/chef/mixin/from_file.rb:42:in `class_eval'
c:/opscode/chef/embedded/apps/chef/lib/chef/mixin/from_file.rb:42:in `class_from_file'
c:/opscode/chef/embedded/apps/chef/lib/chef/resource/lwrp_base.rb:52:in `build_from_file'
c:/chef/runs/2c93403f-afae-4c00-ba35-1579d9b11645/local-mode-cache/cache/cookbooks/compat_resource/files/lib/chef_compa
t/monkeypatches/chef/resource/lwrp_base.rb:50:in `build_from_file'
c:/opscode/chef/embedded/apps/chef/lib/chef/run_context/cookbook_compiler.rb:220:in `load_lwrp_resource'
c:/opscode/chef/embedded/apps/chef/lib/chef/run_context/cookbook_compiler.rb:205:in `block in load_lwrps_from_cookbook'
c:/opscode/chef/embedded/apps/chef/lib/chef/run_context/cookbook_compiler.rb:204:in `each'
c:/opscode/chef/embedded/apps/chef/lib/chef/run_context/cookbook_compiler.rb:204:in `load_lwrps_from_cookbook'
c:/opscode/chef/embedded/apps/chef/lib/chef/run_context/cookbook_compiler.rb:121:in `block in compile_lwrps'
c:/opscode/chef/embedded/apps/chef/lib/chef/run_context/cookbook_compiler.rb:120:in `each'
c:/opscode/chef/embedded/apps/chef/lib/chef/run_context/cookbook_compiler.rb:120:in `compile_lwrps'
c:/opscode/chef/embedded/apps/chef/lib/chef/run_context/cookbook_compiler.rb:73:in `compile'
c:/opscode/chef/embedded/apps/chef/lib/chef/run_context.rb:96:in `load'
c:/opscode/chef/embedded/apps/chef/lib/chef/policy_builder/expand_node_object.rb:73:in `setup_run_context'
c:/opscode/chef/embedded/apps/chef/lib/chef/client.rb:249:in `setup_run_context'
c:/opscode/chef/embedded/apps/chef/lib/chef/client.rb:446:in `run'
c:/opscode/chef/embedded/apps/chef/lib/chef/application.rb:253:in `run_with_graceful_exit_option'
c:/opscode/chef/embedded/apps/chef/lib/chef/application.rb:230:in `block in run_chef_client'
c:/opscode/chef/embedded/apps/chef/lib/chef/local_mode.rb:38:in `with_server_connectivity'
c:/opscode/chef/embedded/apps/chef/lib/chef/application.rb:213:in `run_chef_client'
c:/opscode/chef/embedded/apps/chef/lib/chef/application/client.rb:392:in `block in interval_run_chef_client'
c:/opscode/chef/embedded/apps/chef/lib/chef/application/client.rb:382:in `loop'
c:/opscode/chef/embedded/apps/chef/lib/chef/application/client.rb:382:in `interval_run_chef_client'
c:/opscode/chef/embedded/apps/chef/lib/chef/application/client.rb:372:in `run_application'
c:/opscode/chef/embedded/apps/chef/lib/chef/application.rb:60:in `run'
c:/opscode/chef/embedded/apps/chef/bin/chef-client:26:in `<top (required)>'
c:/opscode/chef/bin/chef-client:63:in `load'
c:/opscode/chef/bin/chef-client:63:in `<main>'

Whatever compat_resource translation that's supposed to be going on to enable newer custom resources to work under earlier versions of Chef doesn't appear to work.
